File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_PING.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.IO;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_PING : Message
    {
        public SID_PING()
        {
            Id = (byte)MessageIds.SID_PING;
            Buffer = new byte[0];
        }

        public SID_PING(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_PING;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            if (context.Client == null || !context.Client.Connected || context.Client.GameState == null) return false;
            var gameState = context.Client.GameState;

            switch (context.Direction)
            {
                case MessageDirection.ClientToServer:
                {
                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");

                    if (Buffer.Length != 4)
                        throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be 4 bytes");

                    using var m = new MemoryStream(Buffer);
                    using var r = new BinaryReader(m);
                    var token = r.ReadUInt32();

                    gameState.LastPong = DateTime.Now;
                    var delta = gameState.LastPong - gameState.LastPing;

                    var autoRefreshPings = Settings.GetBoolean(new string[] { "battlenet", "emulation", "auto_refresh_pings" }, false);
                    if (gameState.Ping != -1 && !autoRefreshPings) return true;

                    if (gameState.ActiveChannel == null)
                        gameState.Ping = (int)Math.Round(delta.TotalMilliseconds);
                    else
                        gameState.ActiveChannel.UpdateUser(gameState, gameState.Ping);

                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Ping: {gameState.Ping}ms");
                    return true;
                }
                case MessageDirection.ServerToClient:
                {
                    var token = (UInt32)(context.Arguments.ContainsKey("token") ? context.Arguments["token"] : (new Random()).Next());

                    Buffer = new byte[4];
                    using var m = new MemoryStream(Buffer);
                    using var w = new BinaryWriter(m);
                    w.Write((UInt32)token);

                    gameState.LastPing = DateTime.Now;
                    gameState.PingToken = token;

                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");
                    context.Client.Send(ToByteArray(context.Client.ProtocolType));
                    return true;
                }
                default: return false;
            }
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_PING.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.IO

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_PING
        Inherits Message

        Public Sub New()
            Id = CByte(MessageIds.SID_PING)
            Buffer = Array.Empty(Of Byte)()
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_PING)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")

            If context.Arguments IsNot Nothing AndAlso context.Arguments.ContainsKey("token") Then
                Dim t = CUInt(context.Arguments("token"))
                Buffer = New Byte(3) {}

                Using _m = New MemoryStream(Buffer)
                    Using _w = New BinaryWriter(_m)
                        _w.Write(t)
                    End Using
                End Using
            End If

            If Buffer.Length <> 4 Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be 4 bytes")
            If context.Client Is Nothing OrElse Not context.Client.Connected Then Return False
            Dim delta = DateTime.Now - context.Client.GameState.PingDelta
            Dim autoRefreshPings = Settings.GetBoolean(New String() {"battlenet", "emulation", "auto_refresh_pings"}, False)
            If Not autoRefreshPings AndAlso context.Client.GameState.Ping <> -1 Then Return True

            Dim token As UInt32
            Using m = New MemoryStream(Buffer)
                Using r = New BinaryReader(m)
                    token = r.ReadUInt32()
                End Using
            End Using

            If Not (context.Direction = MessageDirection.ClientToServer AndAlso token = context.Client.GameState.PingToken) Then
                Return True
            End If
            context.Client.GameState.Ping = CInt(Math.Round(delta.TotalMilliseconds))
            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Ping: {context.Client.GameState.Ping}ms")

            If context.Client.GameState.ActiveChannel IsNot Nothing Then
                context.Client.GameState.ActiveChannel.UpdateUser(context.Client.GameState, context.Client.GameState.Ping)
            End If

            Return True
        End Function
    End Class
End Namespace
